signbank/query_parameters.py

```diff
@@ -493,3 +494,503 @@ def pretty_print_query_values(dataset_languages,query_parameters):
             query_dict[keyword_field_name] = query_parameters[keyword_field_name]
 
     return query_dict
+
+
+def search_fields_from_get(searchform, GET):
+    """
+    Collect non-empty search fields from GET into dictionary
+    Called from get_context_data
+    :form: MorphemeSearchForm
+    :view: MorphemeListView
+    :model: Morpheme
+    """
+    search_keys = []
+    search_fields_to_populate = dict()
+    if not searchform:
+        return search_keys, search_fields_to_populate
+    search_form_fields = searchform.fields.keys()
+    for get_key, get_value in GET.items():
+        if get_value in ['', '0']:
+            continue
+        if get_key.endswith('[]'):
+            vals = GET.getlist(get_key)
+            search_fields_to_populate[get_key] = vals
+            search_keys.append(get_key)
+        elif get_key not in search_form_fields:
+            # skip csrf_token and page
+            continue
+        elif get_key in ['search', 'translation']:
+            from signbank.tools import strip_control_characters
+            val = strip_control_characters(get_value)
+            search_fields_to_populate[get_key] = escape(val)
+            search_keys.append(get_key)
+        else:
+            search_fields_to_populate[get_key] = get_value
+            search_keys.append(get_key)
+
+    return search_keys, search_fields_to_populate
```

Comment:
@susanodd Can't you just remove `search_keys` and just do `search_fields_to_populate.keys()` instead, preferably where this function is called, or even in the relevant template?

Reply:
See the comment below. Somehow if a dictionary is empty, doing dictionary.keys() gives an error.

Reply:
Perhaps it needs to be cast to a list somewhere:

```python
>>> {}.keys()
dict_keys([])
>>> dict().keys()
dict_keys([])
>>> list({}.keys())
[]
```